homeassistant/components/yamaha_musiccast/config_flow.py

```diff
@@ -82,16 +82,18 @@ def _show_setup_form(
             errors=errors or {},
         )
 
-    async def async_step_ssdp(self, discovery_info) -> data_entry_flow.FlowResult:
+    async def async_step_ssdp(
+        self, discovery_info: ssdp.SsdpServiceInfo
+    ) -> data_entry_flow.FlowResult:
         """Handle ssdp discoveries."""
         if not await MusicCastDevice.check_yamaha_ssdp(
-            discovery_info[ssdp.ATTR_SSDP_LOCATION], async_get_clientsession(self.hass)
+            discovery_info.ssdp_location, async_get_clientsession(self.hass)
         ):
             return self.async_abort(reason="yxc_control_url_missing")
 
-        self.serial_number = discovery_info[ssdp.ATTR_UPNP_SERIAL]
-        self.host = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION]).hostname
-        self.upnp_description = discovery_info[ssdp.ATTR_SSDP_LOCATION]
+        self.serial_number = discovery_info.upnp[ssdp.ATTR_UPNP_SERIAL]
+        self.host = urlparse(discovery_info.ssdp_location or "").hostname or ""
```

Comment:
Does this mean we can update the config entry host to be an empty string? Is that valid?

Reply:
Hi @MartinHjelmare
This change was to satisfy `mypy`:
- `urlparse` reacts differently if it is passed `str` versus `str | None`
- then the generated `hostname` can still be None

This should be caught on line 87 though, so maybe using the form `if TYPE_CHECKING` twice (once for `ssdp_location`, and then again for `hostname`) would have been better?
https://github.com/epenet/home-assistant-core/blob/3b0d984959beef06a29a620815de4f3e810c73e5/homeassistant/components/yamaha_musiccast/config_flow.py#L87

I think using the standard form `if discovery_info.ssdp_location is None` it would have become impossible to build a test for it.

Reply:
I can refactor as part of #60787 if needed.

Reply:
For the discussion around `urlparse`, see this comment from a previous `hue` PR: https://github.com/home-assistant/core/pull/60598#discussion_r759258314

Reply:
We shouldn't accommodate bugs or write code that looks like a bug to satisfy typing. Either use `if TYPE_CHECKING` or some other explicit way of setting the type without allowing a potential bug to sneak in.